### PR TITLE
General unroller

### PR DESCRIPTION
**Pull request by jnice-81**
Unrolled scheduler now supports unrolling maps anywhere in the SDFG, also if they contain nested SDFG's.
Adds 2 tests, one that checks nested unrolling with nested SDFG's, and one much simpler test, that unrolls a map, containing one tasklet.
The general concept of unrolling is to backup all the fields that might be affected by calls to replace, then replacing all the map parameters and generating the scope, followed by restoring the fields that were saved.

### CODE REVIEW

File: dace/sdfg/replace.py
Comment (definelicht) [inline dace/sdfg/replace.py:29]: I'm a bit uncomfortable with this: The reason why we don't normally traverse into nested SDFGs when doing replacement is that symbols don't necessarily mean the same thing, `i` could refer to one thing in one SDFG, and something else in another.

Why did you need this? Perhaps we can find another solution, using `sdfg.constants` and `nested_sdfg.symbol_mapping` to solve this without doing the replacement in nested SDFGs? Or, at least, we should make sure that the symbol _does_ mean the same thing (by checking that it is not defined again in the nested SDFG and that the symbol mapping doesn't map it to something else).

File: tests/codegen/unroller_general_test.py
Comment (definelicht) [inline tests/codegen/unroller_general_test.py:77]:
```suggestion
```

File: tests/codegen/unroller_general_test.py
Comment (definelicht) [inline tests/codegen/unroller_general_test.py:99]:
```suggestion
```

File: tests/codegen/unroller_general_test.py
Comment (definelicht) [inline tests/codegen/unroller_general_test.py:82]: What's on here, you're assigning some values after running the program?

File: tests/codegen/unroller_general_test.py
Comment (definelicht) [inline tests/codegen/unroller_general_test.py:74]:
```suggestion
```

No need, it's easy to just check `.dacecache` and open the relevant code file, don't need to print it to the terminal :-)

File: dace/codegen/targets/unroller.py
Comment (definelicht) [inline dace/codegen/targets/unroller.py:20]: Could you explain why we need these backup functions? We can we not just deep copy, change the deep-copied graph, then pass this to the dispatcher?

**Review by definelicht — CHANGES_REQUESTED** [review]

File: tests/codegen/unroller_general_test.py
Comment (jnice-81) [reply tests/codegen/unroller_general_test.py:82]: That's just convenience, since the sdfg is supposed to return exactly the same values that were passed with the only exception of the values at [1, 0], [0, 0]. In other words this generates the test values that should be returned. (Maybe I should copy it to a second variable for the sake of clarity)

File: dace/codegen/targets/unroller.py
Comment (jnice-81) [reply dace/codegen/targets/unroller.py:20]: The reason for this is that deepcopy doesn't work well with subscopes of the graph. There are functions and parts of the codegen, that access state data on the top-level (outside of the copied scope, e.g. on state or sdfg-level) which means you can still get the old objects for some type of calls. This then leads to all sorts of weird behavior.

**Comment by jnice-81** [comment]
Do the failing tests here mean that there is something significantly wrong with this pull request?

File: dace/sdfg/replace.py
Comment (jnice-81) [reply dace/sdfg/replace.py:29]: Assuming I would just add the value to sdfg.constants before generating it and afterwards removing them again, would this behave as expected? (I.e. use the constant if it's not defined, otherwise take whatever it is in the nested sdfgs)

**Comment by definelicht** [comment]
> Do the failing tests here mean that there is something significantly wrong with this pull request?

Yes. You should be able to click the failing tests and see the errors. Hopefully they are all within the code you added, but if not, feel free to ask for advice on what went wrong.

File: dace/codegen/targets/unroller.py
Comment (definelicht) [reply dace/codegen/targets/unroller.py:20]: Ok, I understand. As usual, I'm not super comfortable with this, but also don't have an amazing alternative idea.

The only idea I have is whether we could limit the "backup" to be part of the replacement traversal only: When `replace` traverses the SDFG and finds an occurrences of the token that we are replacing, save a backup before replacing it, and return a tuple of replaces entities (similar to what you are doing now). Then, a separate "unreplace" ("replace inversion"? "rereplace"?) understands this format and reverts these changes. The advantage over what you are already doing is that we _only backup what we change_, so we don't risk backing up/restoring too much or too little.

Does this make any sense?

File: dace/sdfg/replace.py
Comment (definelicht) [reply dace/sdfg/replace.py:29]: It _should_ work with nested SDFGs, I'm just saying that we need to be more careful there because of the name clash I mentioned above. If you fix the variable `w` in the outer SDFG, and then do a symbol mapping of, e.g., `{"w": "w"}` in the nested SDFG, does this still not produce a constant?

If not, then I would prefer the solution of adding them to `sdfg.constants` as you suggested.

Comment (jnice-81) [reply dace/sdfg/replace.py]: I already implemented it now with constants. Should be fine. Copied from http://candcplusplus.com/c-variables-scopelocal-variables-and-global-variables:
"If a global and local variable have the same name and if we try to access the variable then note the local scope variable with the same name has a greater influence in that local region and so it will be used.However,if the local variable cease to exist then the global variable will be used instead.Consider the example below." Since dace always creates local regions whenever a variable is created, it should work out, but I will create a little addition to the test to verify it

File: dace/codegen/targets/unroller.py
Comment (jnice-81) [reply dace/codegen/targets/unroller.py:17]: Another suggestion: I'm not so sure if it is even so clever to change the actual sdfg during unrolling. In principle, we could do the same thing as with the constants in sdfgs, i.ee we create a local scope for each unrolled section (I believe thats already done in face), and then define a constexpr before each section? The only bad thing I could see is that something like nested unrolling would not work You could still have nested maps with unrolling, but you may not bind the range of the inner map to the variable defined by the outer map.

Comment (definelicht) [reply dace/codegen/targets/unroller.py:17]: If this works then yes absolutely, I'm not sure about Intel though, since OpenCL doesn't support `constexpr`. Do you want to try?

Comment (definelicht) [reply dace/codegen/targets/unroller.py:17]: It's definitely not as simple, I doubt you can even do shadowing of OpenCL constant variables. Honestly you're probably better off using `#define` and `#undef` or something nasty like this.

> how can I find out from the unroller if it is currently generating Intel FPGA code?

The dispatcher calls specific code generators, which then call other functions. Some of these other functions take the code generator currently running as an argument (typically a keyword argument `codegen`), but not necessarily all of them. In this case I don't see any easy way for you to know, since this is implemented as its own code generator, so by definition it's neither Xilinx or Intel.

Honestly, we might be better off holding off on Intel OpenCL support. If you can make it work for Xilinx and CPU, then we can leave this pain for another time. @TizianoDeMatteis what do you think?

File: dace/sdfg/replace.py
Comment (definelicht) [reply dace/sdfg/replace.py]: So now we got rid of `deepreplace`? Did we even get rid of the backup stuff as well?

File: dace/codegen/targets/unroller.py
Comment (TizianoDeMatteis) [reply dace/codegen/targets/unroller.py:17]: Constants have always been problematic for Intel OpenCL: you can't do shadowing, and, we currently use`#define` to handle some strange situations. I'm also in favor of the `TODO(later)`

Comment (jnice-81) [reply dace/codegen/targets/unroller.py:17]: alright then I'l just do constexpr, add a TODO and get rid of the replacement/backup stuff

**Comment by jnice-81** [comment]
Changes would again be ready to review. I play the a little bit dirty trick of adding the variables to the sdfg constants for the duration of the scope generation, even though they are actually not. This is required for HBM, because the magic index has to be evaluated during codegen.

File: dace/codegen/targets/unroller.py
Comment (definelicht) [inline dace/codegen/targets/unroller.py:81]: Would it work to just always set _both_ `name` and `unique_name`, regardless of the configuration? It would clean up this whole function quite a lot

File: dace/codegen/targets/unroller.py
Comment (definelicht) [inline dace/codegen/targets/unroller.py:110]:
```suggestion
```

File: dace/codegen/targets/unroller.py
Comment (definelicht) [inline dace/codegen/targets/unroller.py:139]:
```suggestion
```

File: dace/codegen/targets/unroller.py
Comment (definelicht) [inline dace/codegen/targets/unroller.py:123]: Use f-string formatting:

```python
f"constexpr long long {param} {dace.codegen.targets.common.sym2cpp(index)}"
```

File: tests/codegen/unroller_general_test.py
Comment (definelicht) [inline tests/codegen/unroller_general_test.py:75]:
```suggestion
```

**Review by definelicht — CHANGES_REQUESTED** [review]

File: dace/codegen/targets/unroller.py
Comment (jnice-81) [reply dace/codegen/targets/unroller.py:81]: Theoretically yes, but I'm a bit afraid to do so, because I'm not so sure how unique_name is supposed to be used. However at the moment the only place that uses this at all is the nested sdfg generation, so it's probably fine.

**Comment by jnice-81** [comment]
Again everything updated.

**Review by definelicht — APPROVED** [review]

**Comment by definelicht** [comment]
Looks good for me, but I would just like for Tal to take a look before we merge

**Comment by jnice-81** [comment]
By the way which script do I have to run to check everything in tests/codegen? Respectively which tests can I even run locally before declaring that it works?

**Comment by definelicht** [comment]
I'll trigger the CI. You can test your own test locally, and then if anything else has broken you can debug it locally.

**Review by tbennun — CHANGES_REQUESTED** [review]
Minor comments and a question

File: dace/codegen/targets/unroller.py
Comment (jnice-81) [reply dace/codegen/targets/unroller.py:39]: Yes, both Tiziano and Johannes wanted to implement this at some later point when it is actually needed

File: dace/codegen/targets/unroller.py
Comment (jnice-81) [reply dace/codegen/targets/unroller.py:101]: I'm actually not sure, but I don't think it's a problem, because you anyway have to be in a map to use this schedule type, which means you are inside function-code, with the exception of nested sdfg's, but they are handled seperately.

Comment (jnice-81) [reply dace/codegen/targets/unroller.py:101]: In fact I did not even change that, one curly brace was already there, I just added doing a second one

**Review by definelicht — APPROVED** [review]

**Review by tbennun — APPROVED** [review]

**Pull request merged**
